src/video.cpp

```diff
@@ -5,861 +5,908 @@
+#endif
 
 /**
- * Initialize the video system, preparing it for use.
- * @param font_name Name of the font file to load.
- * @param font_size Size of the font.
- * @return Error message if initialization failed, else an empty text.
+ * Called by OpenGL when the window size changes.
+ * @param window Window whose size changed.
+ * @param w New width.
+ * @param h New height.
  */
-std::string VideoSystem::Initialize(const std::string &font_name, int font_size)
+void VideoSystem::FramebufferSizeCallback(GLFWwindow *window, int w, int h)
 {
-	if (this->initialized) return "";
-
-	if (SDL_Init(SDL_INIT_VIDEO) != 0) {
-		std::string err = "SDL video initialization failed: ";
-		err += SDL_GetError();
-		return err;
-	}
-
-	std::string caption = "FreeRCT ";
-	caption += _freerct_revision;
-	this->window = SDL_CreateWindow(caption.c_str(), SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, 800, 600, SDL_WINDOW_RESIZABLE);
-	if (this->window == nullptr) {
-		std::string err = "SDL window creation failed: ";
-		err += SDL_GetError();
-		SDL_Quit();
-		return err;
-	}
-
-	this->renderer = SDL_CreateRenderer(this->window, -1, SDL_RENDERER_ACCELERATED);
-	if (this->renderer == nullptr) {
-		std::string err = "SDL renderer creation failed: ";
-		err += SDL_GetError();
-		SDL_DestroyWindow(this->window);
-		SDL_Quit();
-		return err;
-	}
-
-	this->GetResolutions();
+	assert(window == _video.window);
+	assert(w >= 0);
+	assert(h >= 0);
 
-	this->SetResolution({800, 600}); // Allocates this->mem, return value is ignored.
+	_video.width = w;
+	_video.height = h;
+	_video.UpdateClip();
 
-	/* SDL_CreateRGBSurfaceFrom() pretends to use a void* for the data,
-	 * but it's really treated as endian-specific uint32*.
-	 * But since the c++ compiler handles the endianness of _icon_data, it's ok. */
-	const uint32 rmask = 0xff000000;
-	const uint32 gmask = 0x00ff0000;
-	const uint32 bmask = 0x0000ff00;
-	const uint32 amask = 0x000000ff;
-	SDL_Surface *icon = SDL_CreateRGBSurfaceFrom((void *)_icon_data, 32, 32, 4 * 8, 4 * 32,
-	                                             rmask, gmask, bmask, amask);
-
-	if (icon != nullptr) {
-		SDL_SetWindowIcon(this->window, icon);
-		SDL_FreeSurface(icon);
-	} else {
-		printf("Could not set window icon (%s)\n", SDL_GetError());
-	}
+	_window_manager.RepositionAllWindows(_video.width, _video.height);
+	NotifyChange(WC_BOTTOM_TOOLBAR, ALL_WINDOWS_OF_TYPE, CHG_RESOLUTION_CHANGED, 0);
+}
 
-	SDL_StartTextInput(); // Enable Unicode character input.
+/**
+ * Called by OpenGL when a key was pressed.
+ * @param window Window in which the event occurred.
+ * @param key Constant of the pressed key.
+ * @param scancode Scancode of the pressed key.
+ * @param action Whether the mouse button was pressed or released.
+ * @param mods Bitset of modifier keys.
+ */
+void VideoSystem::KeyCallback(GLFWwindow *window, int key, [[maybe_unused]] int scancode, int action, int mods)
+{
+	assert(window == _video.window);
+	if (action == GLFW_RELEASE) return;
```

Comment:
Alternatively, you could return if `action != GLFW_PRESS` (I assume). That excludes any other key event that may happen as well.